### PR TITLE
adding client-metadata

### PR DESCRIPTION
**Pull request by Eransho**
Adding the new `client-metadata` argument

### CODE REVIEW

File: src/TabNine.ts [hunk 1]
```diff
@@ -195,5 +203,29 @@ export class TabNine {
...
+}
+
+export class TabNineExtension {
+  private extension: vscode.Extension<any>;
+  
+  private constructor(ext: vscode.Extension<any>) {
+    this.extension = ext;
+  }
+
+  getExtensionPath(): string {
```
Comment (dimacodota) [inline src/TabNine.ts:215]: you can defined them as getters: `get version() {}` and then call them as properties `ext.version`

**Comment by Eransho** [comment]
BTW, is there a reason not to use `vscode.env.appName` as the client string argument instead of the fixed "vscode" string?

File: src/TabNine.ts [hunk 2]
```diff
@@ -87,8 +91,12 @@ export class TabNine {
   }
 
   private static runTabNine(inheritStdio : boolean = false, additionalArgs: string[] = []): child_process.ChildProcess {
+    const ext = TabNineExtension.getInstance();
```
Comment (dimacodota) [inline src/TabNine.ts:94]: maybe we should create this only once? not for each binary start, this is "static" metadata related to the extension that will not change between the binary instances

**Comment by dimacodota** [comment]
> BTW, is there a reason not to use `vscode.env.appName` as the client string argument instead of the fixed "vscode" string?

it will behave differently in vscode-insiders, you will get "Visual Studio Code - Insiders" while in vscode you will get "Visual Studio Code"

**Comment by Eransho** [comment]
> > BTW, is there a reason not to use `vscode.env.appName` as the client string argument instead of the fixed "vscode" string?
> 
> it will behave differently in vscode-insiders, you will get "Visual Studio Code - Insiders" while in vscode you will get "Visual Studio Code"

I assumed it will be a different value. The question is do we want to distinguish between the "different" products

File: src/TabNine.ts (hunk 1 above)
Comment (Eransho) [reply src/TabNine.ts:215]: cool, thanks 👍

File: src/TabNine.ts (hunk 2 above)
Comment (Eransho) [reply src/TabNine.ts:94]: makes sense

File: src/TabNine.ts [hunk 3]
```diff
@@ -87,8 +91,12 @@ export class TabNine {
...
     const args = [
       "--client=vscode",
+      "--client-metadata",
+      "clientVersion=" + vscode.version,
+      "pluginVersion=" + ext.version,
```
Comment (amircodota) [inline src/TabNine.ts:99]: @dimacodota maybe we also one if this is an insiders build or not?

Comment (dimacodota) [reply src/TabNine.ts:99]: we will get this as part of the `vscode.version`, if it is insiders the version will be `-insiders`

**Pull request merged**